src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,237 @@
+    unsafe fn any_as_u8_slice<T: Sized>(p: &T) -> &[u8] {
+        ::std::slice::from_raw_parts(
+            (p as *const T) as *const u8,
+            ::std::mem::size_of::<T>(),
+        )
+    }
+    /// Get precomputed keys
+    /// If the Key is not found in cache, create symmetric key and insert it into cache for later use.
+    pub fn get_symmetric_key (&mut self, key_pair: (&SecretKey, &PublicKey)) -> Option<PrecomputedKey> {
+        let sec_arr = unsafe { PrecomputedKeys::any_as_u8_slice(&key_pair.0) };
+        let key = [key_pair.1.as_ref(), sec_arr].concat();
+        
+        match self.cache.get(&key) {       // if symmetric key exists in cache, returns with the value
+            Some(k) => {
+                return PrecomputedKey::from_slice(k);
+            },
+            None => { ; },
+        };
+
+        // Key don't exist in cache, So create one
+        // must separate logic into two blocks because self.cache is barrowed mutably
+        let shared_secret = &encrypt_precompute(key_pair.1, key_pair.0);
+        let shared_arr = unsafe { PrecomputedKeys::any_as_u8_slice(&shared_secret) };
+        self.cache.insert (key.clone(), shared_arr.to_vec());
+        return Some(shared_secret.clone());
+    }
+}
+
+impl DhtPacket {
+    /// create new DhtPacket object
+    pub fn new(shared_secret: &PrecomputedKey, pk: &PublicKey, dp: DhtPacketPayload) -> DhtPacket {
+        let nonce = &gen_nonce();
+        let mut buf = [0; MAX_DHT_PACKET_SIZE];
+        let payload = seal_precomputed(dp.to_bytes((&mut buf, 0)).ok().unwrap().0, nonce, shared_secret);
+
+        DhtPacket {
+            packet_kind: dp.kind(),
+            pk: *pk,
+            nonce: *nonce,
+            payload: payload,
+        }
+
+    }
+
+    /**
+    Decrypt payload and try to parse it as packet type.
+
+    To get info about it's packet type use
+    [`.kind()`](./struct.DhtPacket.html#method.kind) method.
+
+    Returns `None` in case of faliure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    /* TODO: perhaps switch to using precomputed symmetric key?
+              - given that computing shared key is apparently the most
+                costly operation when it comes to crypto, using precomputed
+                key might (would significantly?) lower resource usage
+
+                Alternatively, another method `get_payloadnm()` which would use
+                symmetric key.
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = match open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key) {
+            Ok(d) => d,
+            Err(_) => {
+                debug!("Decrypting DhtPacket failed!");
+                return Ok(None);
```

Comment:
I don't think that `Ok(None)` is a good idea.

Reply:
I agree.
If the result of this function is Err, it should contain Err info.
I will change this.